lib/vast/Conversion/Common/CoreToLLVM.cpp

```diff
@@ -665,6 +684,73 @@ namespace vast
             }
         };
 
+        template< typename LOp >
+        struct logical : BasePattern< LOp >
+        {
+            using Base = BasePattern< LOp >;
+            using Base::Base;
+
+            mlir::LogicalResult matchAndRewrite(
+                LOp op, typename LOp::Adaptor ops,
+                mlir::ConversionPatternRewriter &rewriter) const override
+            {
+                // TODO: catch edge cases (e.g. operand is a constant)
+                auto lhs_type = this->type_converter().convertType(ops.getLhs().getType());
+
+                // Creates llvm dialect constant 0 for comparison
+                auto create_zero = [&](){
+                    return rewriter.create< LLVM::ConstantOp >(
+                        op.getLoc(), lhs_type, rewriter.getIntegerAttr(lhs_type, 0)
+                    );
+                };
+
+                auto cmp_lhs = rewriter.create< LLVM::ICmpOp >(op.getLoc(),
+                    LLVM::ICmpPredicate::eq,
+                    ops.getLhs(),
+                    create_zero());
+
+                //split the current block for lazy evaluation of rhs
+                auto curr_block = rewriter.getBlock();
+                auto rhs_block = curr_block->splitBlock(op);
+                auto end_block = rhs_block->splitBlock(rhs_block->begin());
+                end_block->addArgument(cmp_lhs.getResult().getType(), op.getLoc());
+
+                rewriter.setInsertionPointToStart(rhs_block);
+                auto cmp_rhs = rewriter.create< LLVM::ICmpOp >(op.getLoc(),
+                    LLVM::ICmpPredicate::eq,
+                    ops.getRhs(),
+                    create_zero());
+                rewriter.create< LLVM::BrOp >(op.getLoc(), cmp_rhs.getResult(), end_block);
+
+                rewriter.setInsertionPointToEnd(curr_block);
+                if (std::same_as< LOp, hl::BinLOrOp >)
```

Comment:
if constexpr

but I would prefer to have two separate functions, or template that define this rewrite